### DISCUSSION TITLE
Is there any example of Wizard customization?

### DISCUSSION THREAD

**Discussion by mshdevtech** [data-driven-forms/react-forms#989] [Need help]
I need to add custom handlers for prev and next buttons.

**Comment by rvsia** [comment]
Hello, what mapper do you use?

The only mapper that currently supports buttons customization in wizard is the PF4 mapper.

We can add it to other mappers as well, if there is a requirement to do it.

(also transfering this to discussion)

**Reply by mshdevtech** [reply]
Hi @rvsia 
I'm using Material UI mapper. Is it possible to fully overwrite wizard and use custom template and flow?

**Reply by rvsia** [reply]
No, not at this moment. However, since the release of version 3, we made [common package](https://www.npmjs.com/package/@data-driven-forms/common) public so you can take the wizard internals and you can easily reimplement the wizard component by yourself. (Just follow the [original component](https://github.com/data-driven-forms/react-forms/blob/master/packages/mui-component-mapper/src/wizard/wizard.js))

However, because it's an internal package and we do not propagate it anywhere there could be theoretically small changes in the future - so it would be better to have the wizard maximally tested (or to follow changes in this package.)

**Reply by mshdevtech** [reply]
Hi @rvsia 
I'm trying to reimplement wizard component. But I have one issue with overriding components.

When I'm doing like this and then use 'my-wizard' in name of custom wizard component:

```
import { merge } from 'lodash-es';
import MuiComponentMapper from '@data-driven-forms/mui-component-mapper/component-mapper';
import MyWizard from 'forms/components/MyWizard';

const componentMapper = merge({}, MuiComponentMapper, {
  'my-wizard': MyWizard,
});

export default componentMapper;
```

It throws me an error:

> Form could not be rendered, because of invalid form schema.
> DefaultSchemaError:
> Each fields item must have "component" property!
> If you don't know what this error means, please contact site administrator.

But this works fine:

```
import { merge } from 'lodash-es';
import MuiComponentMapper from '@data-driven-forms/mui-component-mapper/component-mapper';
import MyWizard from 'forms/components/MyWizard';

const componentMapper = merge({}, MuiComponentMapper, {
  wizard: MyWizard,
});

export default componentMapper;
```

**Reply by rvsia** [reply]
@maxdev oh, well. There is an exception that wizard steps do not have to have any component attribute and this exception is bind to the `wizard` keyword. I think the overwriting is correct, if you don’t need to use both wizard components at the same time in one form. Or you can add a custom fake “wizard-step” component and use it in the schema in wizard step fields to pass the schema validation.

**Reply by mshdevtech** [reply]
Thanks for reply. I just was confused if I'm doing in right way. Luckily I don't need different wizards. But I suppose it would be nice to have few words about this in documentation. 😉

**Reply by rvsia** [reply]
Yeah, I agree, but unfortunately it hasn't been a big issue before and so it has a very low priority. (Actually, I wanted always to make it more configurable.)

**Reply by Hyperkid123** [reply]
@maxdev If you end up re-implementing the wizard and it will use the common package, could you share it with us? I think it would be a great addition.
